mainmatter/background_and_related_work/related_work/index.tex

```diff
@@ -1,82 +1,105 @@
-\todo{We always have to write a lot of shaders, the problem is that there's no support for scaling up}
 
-\item Different semantics for each shader stage. 
-This contributes to the issue of having to keep track of semantics depending on what kind of shader is being worked on.
-\todo{Give an example of how semantics are mismatched}
-It makes it more difficult to use OpenGL, and increases the learning curve. 
+The shaders in \cref{fig:openglShaders} communicate using the fragPos variable, but if this name does not match in the two programs, we are only informed of this error at runtime. 
+While the shaders in \cref{fig:openglShaders} are rather simple, a graphics application often needs many different types of shaders. To make the use of shaders flexible, metaprogramming is used to combine snippets of \gls{GLSL} code as to construct the needed shaders. This is done using the \#define and \#ifdef directives, but as the amount of shaders needed scales up the programs created becomes unmanageable.
 
-\item No type system for vectors when converting between spaces. 
-\todo{We have to remember to talk about spaces}
-It has to be done manually, it should be possible to simply standardize such an approach.
 
-\item Diffcult to verify correctness of a graphics application. 
-There is no way to test whether whatever an application renders, is the actual desired result.
-\end{enumerate}
+The different kinds of shader stages have different execution semantics, which can make it difficult to understand how the stages relate to each other.
+For instance, the fragment shader runs multiple times for each vertex shader invocation.
+In addition, values passed between the two stages are interpolated during rasterization.
+In the case of \cref{fig:openglShaders}, the fragPos variable will include the vertex position in the vertexShader, but fragPos in the fragment shader represents the interpolation of several vertex positions.
 
-In the conclusion, the article mentions Vulkan and that it might be the solution to some of the issues in OpenGL. 
-Additionally, it also encourages development of new frameworks to rival OpenGL. 
-However a shortcomming of the article is that all the listed issues is only based on the authors opinion that he has discussed with his collegues, and not used a proper methodology to collect this data. 
-Although they do reveal that OpenGL is not a \gls{API} without flaws.
+\fig{figures/openglLighting.pdf}{Excerpt from vertex shader implementing the phong lighting model. \cite{fix_opengl}}{lighting}{1}
 
-The value of this article is that it gives insight as to how a good graphics \gls{API} should work.
-These are good arguments to take into consideration when trying to evaluate an \gls{API}.
+The vertex shader by itself could also benefit from better static type checking. In this shader we convert points in model space to world space, then to camera space and finally to projection space.
```

Comment:
Unless we already mentioned how spaces work, maybe we should refresh what it is. (Bent: Ville i kunne forstå det her hvis i læste det tilbage for 3 måneder siden?)